### ISSUE TITLE
Elastic Search 5 Indexing Performance Issue

### ISSUE THREAD

**Issue by DarthFly** [magento/magento2#20966]
### Preconditions (*)

1. Magento 2.3 and Elastic Search v5 configured.
2. Ability to debug ProductDataMapper.php file - https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php

### Steps to reproduce (*)

1. Create a searchable product attribute with a lot of values. Most common - `brands`. We had around 300 values.
2. Install a lot of products that use values inside this attribute. In our case it was ~180k, but for the debug you may use sample data.
3. Run reindex `bin/magento indexer:reindex catalogsearch_fulltext` (or trigger it in a way to be able to debug)

### Expected result (*)

1. Index is running fine and take sane amount of time.
2.

### Actual result (*)

1. (In our case and develop PC) with elastic 2.3 reindex takes ~40m to complete. With Elastic5 - it was complete in `8h`

Issue comes from this method.
```    
private function getValuesLabels(Attribute $attribute, array $attributeValues): array
    {
        $attributeLabels = [];
        foreach ($attribute->getOptions() as $option) {
            if (\in_array($option->getValue(), $attributeValues)) {
                $attributeLabels[] = $option->getLabel();
            }
        }
        return $attributeLabels;
    }
```
For each product magento run this code providing an array of attribute ids here. This is used for both multiple and single (select) attributes, so `$attributeValues` may look like `[123, 456]` (simplified, value looks a little differently), but for `brands` it mostly contain one value.
`$attribute->getOptions()` return an array with 300+ values, each is compared to the `$attributeValues` array to be able to retrieve a label value for this specific attribute. 

Even if one search takes milliseconds (well, mostly hundreds of milliseconds) multiply this code by amount of products and you will see drop down of performance here by 5 hours.

There are 2 ways to improve this `foreach` fast - cache attribute values by id somewhere and validate if each value inside `$attributeValues` exists as key or cache values that are returned after function (will require joining array). Class is used for every product, so pre-cached value will be used correctly without recalculating it on each product.

**Comment by magento-engcom-team** [comment]
Hi @DarthFly. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@DarthFly do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @DarthFly thank you for your report, which version of elastic you use ?

**Comment by DarthFly** [comment]
@engcom-backlog-nazar Elastic 5 latest and the one installed on cloud. It doesn't really matter as problem is inside magento code, not when data is transferred to and from Elastic.

**Comment by ghost** [comment]
@DarthFly in your case seems like not) but really which is version 5.2.x?
only 5.2.x is supported

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98054`, `MAGETWO-98055` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by DarthFly** [comment]
It any case it was `elasticsearch-5.2.2`

**Comment by ghost** [comment]
@DarthFly thanks, now clear )

**Comment by magento-engcom-team** [comment]
Hi @GovindaSharma. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by jeffminor** [comment]
This is a HUGE issue for us as well. We have 2 store views with 480,xxx skus. Indexing of catalogsearch_fulltext for elasticsearch 5.2.2 went from approximately 60 minutes on 2.2.5 (enterprise edition) to 89+ hours on 2.3. We cannot go to production with this. How could this have been released (especially for enterprise customers) with this large an issue in it?

**Comment by VladimirZaets** [comment]
Hi @DarthFly. Thank you for your report. 
The issue has been fixed in magento/magento2#25452 by @behnamshayani in 2.4-develop branch
Related commit(s):
 - [df506928967e283ce7b976ab42216ee3cff958ab](https://github.com/magento/magento2/commit/df506928967e283ce7b976ab42216ee3cff958ab)
 - [1a0cae28c1c84adbd428cba2e855efa1eab91d4f](https://github.com/magento/magento2/commit/1a0cae28c1c84adbd428cba2e855efa1eab91d4f)
 - [d1c848f39a3811479bd440f9904c0b75a9794f24](https://github.com/magento/magento2/commit/d1c848f39a3811479bd440f9904c0b75a9794f24)
 - [0bf7a4c1b87cf0a738ab17ad76a4cb5b46443167](https://github.com/magento/magento2/commit/0bf7a4c1b87cf0a738ab17ad76a4cb5b46443167)
 - [2b5b9860c395ae85e6c1dc4a736571c9b8985e55](https://github.com/magento/magento2/commit/2b5b9860c395ae85e6c1dc4a736571c9b8985e55)

The fix will be available with the upcoming 2.4.0 release.
